@Gauge in ReqeustScope breaks all metrics

Following the example from [TomiTribe](https://www.tomitribe.com/blog/getting-started-with-microprofile-metrics/#:~:text=The%20MicroProfile%20Metrics%20is%20a,build%20their%20own%20application%20metrics.) we have:

```
@Path("gauge")
@RequestScoped
public class GaugeReproducer {
    @GET
    @Path("/price")
    public Response getItemsPrice() {
        return Response.ok(getPrice()).build();
    }

    @Gauge(unit = "Heads", name = "addHeadGauge", absolute = true)
    public long getPrice() {
        return 4;
    }
}
```
And the request works fine:
```
    public static void main(String[] args) {
        try (Response response = ClientBuilder.newClient()
                .target("http://localhost:8080").path("gauge/price")
                .request(MediaType.APPLICATION_JSON_TYPE).get()) {
            Assertions.assertEquals(200, response.getStatus());
        }
    }
```
But when we made a request to metrics url _http://localhost:8080/metrics/application_ we get a response:
> java.lang.reflect.InvocationTargetException

And the server log contains the full stack trace:
> 2020.06.11 11:17:41 WARNING io.helidon.webserver.RequestRouting Thread[nioEventLoopGroup-3-3,10,main]: Default error handler: Unhandled exception encountered.
java.util.concurrent.ExecutionException: Unhandled 'cause' of this exception encountered.
	at io.helidon.webserver.RequestRouting$RoutedRequest.defaultHandler(RequestRouting.java:387)
	at io.helidon.webserver.RequestRouting$RoutedRequest.nextNoCheck(RequestRouting.java:367)
...
Caused by: java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at io.helidon.microprofile.metrics.DelegatingGauge.getValue(DelegatingGauge.java:66)
	at io.helidon.metrics.HelidonGauge.getValue(HelidonGauge.java:57)
...
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.helidon.microprofile.metrics.DelegatingGauge.getValue(DelegatingGauge.java:64)
...
Caused by: org.jboss.weld.contexts.ContextNotActiveException: WELD-001303: No active contexts for scope type javax.enterprise.context.RequestScoped
	at org.jboss.weld.manager.BeanManagerImpl.getContext(BeanManagerImpl.java:647)

--- a/microprofile/metrics/src/main/java/io/helidon/microprofile/metrics/MetricsCdiExtension.java
+++ b/microprofile/metrics/src/main/java/io/helidon/microprofile/metrics/MetricsCdiExtension.java
@@ -33,6 +33,8 @@ import java.util.Set;
 import java.util.logging.Level;
 import java.util.logging.Logger;
 
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.context.RequestScoped;
 import javax.enterprise.event.Observes;
 import javax.enterprise.inject.Default;
 import javax.enterprise.inject.spi.AfterDeploymentValidation;
@@ -41,6 +43,7 @@ import javax.enterprise.inject.spi.AnnotatedType;
 import javax.enterprise.inject.spi.Bean;
 import javax.enterprise.inject.spi.BeanManager;
 import javax.enterprise.inject.spi.BeforeBeanDiscovery;
+import javax.enterprise.inject.spi.DeploymentException;
 import javax.enterprise.inject.spi.Extension;
 import javax.enterprise.inject.spi.ProcessAnnotatedType;
 import javax.enterprise.inject.spi.ProcessInjectionPoint;
@@ -50,8 +53,10 @@ import javax.enterprise.inject.spi.WithAnnotations;
 import javax.enterprise.inject.spi.configurator.AnnotatedMethodConfigurator;
 import javax.enterprise.inject.spi.configurator.AnnotatedTypeConfigurator;
 import javax.inject.Qualifier;
+import javax.inject.Singleton;
 import javax.interceptor.Interceptor;
 
+import io.helidon.common.Errors;
 import io.helidon.common.HelidonFeatures;
 import io.helidon.common.HelidonFlavor;
 import io.helidon.config.Config;
@@ -98,6 +103,8 @@ public class MetricsCdiExtension implements Extension {
         HelidonFeatures.register(HelidonFlavor.MP, "Metrics");
     }
 
+    private Errors.Collector errors = Errors.collector();
+
     @SuppressWarnings("unchecked")
     private static <T> T getReference(BeanManager bm, Type type, Bean<?> bean) {
         return (T) bm.getReference(bean, type, bm.createCreationalContext(bean));
@@ -124,13 +131,13 @@ public class MetricsCdiExtension implements Extension {
                                               counted.absolute());
             String displayName = counted.displayName().trim();
             Metadata meta = Metadata.builder()
-                .withName(metricName)
-                .withDisplayName(displayName.isEmpty() ? metricName : displayName)
-                .withDescription(counted.description().trim())
-                .withType(MetricType.COUNTER)
-                .withUnit(counted.unit().trim())
-                .reusable(counted.reusable())
-                .build();
+                    .withName(metricName)
+                    .withDisplayName(displayName.isEmpty() ? metricName : displayName)
+                    .withDescription(counted.description().trim())
+                    .withType(MetricType.COUNTER)
+                    .withUnit(counted.unit().trim())
+                    .reusable(counted.reusable())
+                    .build();
             registry.counter(meta, tags(counted.tags()));
             LOGGER.log(Level.FINE, () -> "Registered counter " + metricName);
         } else if (annotation instanceof Metered) {
@@ -139,13 +146,13 @@ public class MetricsCdiExtension implements Extension {
                                               metered.absolute());
             String displayName = metered.displayName().trim();
             Metadata meta = Metadata.builder()
-                .withName(metricName)
-                .withDisplayName(displayName.isEmpty() ? metricName : displayName)
-                .withDescription(metered.description().trim())
-                .withType(MetricType.METERED)
-                .withUnit(metered.unit().trim())
-                .reusable(metered.reusable())
-                .build();
+                    .withName(metricName)
+                    .withDisplayName(displayName.isEmpty() ? metricName : displayName)
+                    .withDescription(metered.description().trim())
+                    .withType(MetricType.METERED)
+                    .withUnit(metered.unit().trim())
+                    .reusable(metered.reusable())
+                    .build();
             registry.meter(meta, tags(metered.tags()));
             LOGGER.log(Level.FINE, () -> "Registered meter " + metricName);
         } else if (annotation instanceof Timed) {
@@ -154,28 +161,28 @@ public class MetricsCdiExtension implements Extension {
                                               timed.absolute());
             String displayName = timed.displayName().trim();
             Metadata meta = Metadata.builder()
-                .withName(metricName)
-                .withDisplayName(displayName.isEmpty() ? metricName : displayName)
-                .withDescription(timed.description().trim())
-                .withType(MetricType.TIMER)
-                .withUnit(timed.unit().trim())
-                .reusable(timed.reusable())
-                .build();
+                    .withName(metricName)
+                    .withDisplayName(displayName.isEmpty() ? metricName : displayName)
+                    .withDescription(timed.description().trim())
+                    .withType(MetricType.TIMER)
+                    .withUnit(timed.unit().trim())
+                    .reusable(timed.reusable())
+                    .build();
             registry.timer(meta, tags(timed.tags()));
             LOGGER.log(Level.FINE, () -> "Registered timer " + metricName);
         } else if (annotation instanceof ConcurrentGauge) {
             ConcurrentGauge concurrentGauge = (ConcurrentGauge) annotation;
             String metricName = getMetricName(element, clazz, lookupResult.getType(), concurrentGauge.name().trim(),
-                    concurrentGauge.absolute());
+                                              concurrentGauge.absolute());
             String displayName = concurrentGauge.displayName().trim();
             Metadata meta = Metadata.builder()
-                .withName(metricName)
-                .withDisplayName(displayName.isEmpty() ? metricName : displayName)
-                .withDescription(concurrentGauge.description().trim())
-                .withType(MetricType.CONCURRENT_GAUGE)
-                .withUnit(concurrentGauge.unit().trim())
-                .reusable(concurrentGauge.reusable())
-                .build();
+                    .withName(metricName)
+                    .withDisplayName(displayName.isEmpty() ? metricName : displayName)
+                    .withDescription(concurrentGauge.description().trim())
+                    .withType(MetricType.CONCURRENT_GAUGE)
+                    .withUnit(concurrentGauge.unit().trim())
+                    .reusable(concurrentGauge.reusable())
+                    .build();
             registry.concurrentGauge(meta, tags(concurrentGauge.tags()));
             LOGGER.log(Level.FINE, () -> "Registered concurrent gauge " + metricName);
         }
@@ -208,7 +215,7 @@ public class MetricsCdiExtension implements Extension {
         return result;
     }
 
-    private static MetricRegistry getMetricRegistry() {
+    static MetricRegistry getMetricRegistry() {
         return RegistryProducer.getDefaultRegistry();
     }
 
@@ -217,7 +224,7 @@ public class MetricsCdiExtension implements Extension {
      *
      * @param discovery bean discovery event
      */
-    public void before(@Observes BeforeBeanDiscovery discovery) {
+    void before(@Observes BeforeBeanDiscovery discovery) {
         LOGGER.log(Level.FINE, () -> "Before bean discovery " + discovery);
 
         // Initialize our implementation
@@ -354,6 +361,12 @@ public class MetricsCdiExtension implements Extension {
             @Observes AfterDeploymentValidation adv, BeanManager bm) {
         LOGGER.log(Level.FINE, () -> "registerProducers");
 
+        Errors problems = errors.collect();
+        errors = null;
+        if (problems.hasFatal()) {
+            throw new DeploymentException("Metrics module found issues with deployment: " + problems.toString());
+        }
+
         MetricRegistry registry = getMetricRegistry();
         producers.entrySet().forEach(entry -> {
             Metric metric = entry.getValue().getAnnotation(Metric.class);
@@ -364,13 +377,13 @@ public class MetricsCdiExtension implements Extension {
                                                   metric.name(), metric.absolute());
                 T instance = getReference(bm, entry.getValue().getBaseType(), entry.getKey());
                 Metadata md = Metadata.builder()
-                    .withName(metricName)
-                    .withDisplayName(metric.displayName())
-                    .withDescription(metric.description())
-                    .withType(getMetricType(instance))
-                    .withUnit(metric.unit())
-                    .reusable(false)
-                    .build();
+                        .withName(metricName)
+                        .withDisplayName(metric.displayName())
+                        .withDescription(metric.description())
+                        .withType(getMetricType(instance))
+                        .withUnit(metric.unit())
+                        .reusable(false)
+                        .build();
                 registry.register(md, instance);
             }
         });
@@ -449,6 +462,21 @@ public class MetricsCdiExtension implements Extension {
             return;
         }
 
+        Annotation annotation = type.getAnnotation(RequestScoped.class);
+        if (annotation != null) {
+            errors.fatal(clazz, "Cannot configure @Gauge on a request scoped bean");
+            return;
+        }
+
+        if (type.getAnnotation(ApplicationScoped.class) == null && type.getAnnotation(Singleton.class) == null) {
+            if (ConfigProvider.getConfig().getOptionalValue("metrics.warn-dependent", Boolean.class).orElse(true)) {
+                LOGGER.warning("@Gauge is configured on a bean " + clazz.getName()
+                                       + " that is neither ApplicationScoped nor Singleton. This is most likely a bug."
+                                       + " You may set 'metrics.warn-dependent' configuration option to 'false' to remove "
+                                       + "this warning.");
+            }
+        }
+
         // Process @Gauge methods keeping non-private declared on this class
         configurator.filterMethods(method -> method.getJavaMember().getDeclaringClass().equals(clazz)
                 && !Modifier.isPrivate(method.getJavaMember().getModifiers())
@@ -484,13 +512,13 @@ public class MetricsCdiExtension implements Extension {
                                           bm);
                 Gauge gaugeAnnotation = site.getAnnotated().getAnnotation(Gauge.class);
                 Metadata md = Metadata.builder()
-                    .withName(gaugeID.getName())
-                    .withDisplayName(gaugeAnnotation.displayName())
-                    .withDescription(gaugeAnnotation.description())
-                    .withType(MetricType.GAUGE)
-                    .withUnit(gaugeAnnotation.unit())
-                    .reusable(false)
-                    .build();
+                        .withName(gaugeID.getName())
+                        .withDisplayName(gaugeAnnotation.displayName())
+                        .withDescription(gaugeAnnotation.description())
+                        .withType(MetricType.GAUGE)
+                        .withUnit(gaugeAnnotation.unit())
+                        .reusable(false)
+                        .build();
                 LOGGER.log(Level.FINE, () -> String.format("Registering gauge with metadata %s", md.toString()));
                 registry.register(md, dg, gaugeID.getTagsAsList().toArray(new Tag[0]));
             } catch (Throwable t) {

--- a/microprofile/metrics/src/main/java/io/helidon/microprofile/metrics/MetricsCdiExtension.java
+++ b/microprofile/metrics/src/main/java/io/helidon/microprofile/metrics/MetricsCdiExtension.java
@@ -131,13 +131,13 @@ public class MetricsCdiExtension implements Extension {
                                               counted.absolute());
             String displayName = counted.displayName().trim();
             Metadata meta = Metadata.builder()
-                    .withName(metricName)
-                    .withDisplayName(displayName.isEmpty() ? metricName : displayName)
-                    .withDescription(counted.description().trim())
-                    .withType(MetricType.COUNTER)
-                    .withUnit(counted.unit().trim())
-                    .reusable(counted.reusable())
-                    .build();
+                .withName(metricName)
+                .withDisplayName(displayName.isEmpty() ? metricName : displayName)
+                .withDescription(counted.description().trim())
+                .withType(MetricType.COUNTER)
+                .withUnit(counted.unit().trim())
+                .reusable(counted.reusable())
+                .build();
             registry.counter(meta, tags(counted.tags()));
             LOGGER.log(Level.FINE, () -> "Registered counter " + metricName);
         } else if (annotation instanceof Metered) {
@@ -146,13 +146,13 @@ public class MetricsCdiExtension implements Extension {
                                               metered.absolute());
             String displayName = metered.displayName().trim();
             Metadata meta = Metadata.builder()
-                    .withName(metricName)
-                    .withDisplayName(displayName.isEmpty() ? metricName : displayName)
-                    .withDescription(metered.description().trim())
-                    .withType(MetricType.METERED)
-                    .withUnit(metered.unit().trim())
-                    .reusable(metered.reusable())
-                    .build();
+                .withName(metricName)
+                .withDisplayName(displayName.isEmpty() ? metricName : displayName)
+                .withDescription(metered.description().trim())
+                .withType(MetricType.METERED)
+                .withUnit(metered.unit().trim())
+                .reusable(metered.reusable())
+                .build();
             registry.meter(meta, tags(metered.tags()));
             LOGGER.log(Level.FINE, () -> "Registered meter " + metricName);
         } else if (annotation instanceof Timed) {
@@ -161,28 +161,28 @@ public class MetricsCdiExtension implements Extension {
                                               timed.absolute());
             String displayName = timed.displayName().trim();
             Metadata meta = Metadata.builder()
-                    .withName(metricName)
-                    .withDisplayName(displayName.isEmpty() ? metricName : displayName)
-                    .withDescription(timed.description().trim())
-                    .withType(MetricType.TIMER)
-                    .withUnit(timed.unit().trim())
-                    .reusable(timed.reusable())
-                    .build();
+                .withName(metricName)
+                .withDisplayName(displayName.isEmpty() ? metricName : displayName)
+                .withDescription(timed.description().trim())
+                .withType(MetricType.TIMER)
+                .withUnit(timed.unit().trim())
+                .reusable(timed.reusable())
+                .build();
             registry.timer(meta, tags(timed.tags()));
             LOGGER.log(Level.FINE, () -> "Registered timer " + metricName);
         } else if (annotation instanceof ConcurrentGauge) {
             ConcurrentGauge concurrentGauge = (ConcurrentGauge) annotation;
             String metricName = getMetricName(element, clazz, lookupResult.getType(), concurrentGauge.name().trim(),
-                                              concurrentGauge.absolute());
+                    concurrentGauge.absolute());
             String displayName = concurrentGauge.displayName().trim();
             Metadata meta = Metadata.builder()
-                    .withName(metricName)
-                    .withDisplayName(displayName.isEmpty() ? metricName : displayName)
-                    .withDescription(concurrentGauge.description().trim())
-                    .withType(MetricType.CONCURRENT_GAUGE)
-                    .withUnit(concurrentGauge.unit().trim())
-                    .reusable(concurrentGauge.reusable())
-                    .build();
+                .withName(metricName)
+                .withDisplayName(displayName.isEmpty() ? metricName : displayName)
+                .withDescription(concurrentGauge.description().trim())
+                .withType(MetricType.CONCURRENT_GAUGE)
+                .withUnit(concurrentGauge.unit().trim())
+                .reusable(concurrentGauge.reusable())
+                .build();
             registry.concurrentGauge(meta, tags(concurrentGauge.tags()));
             LOGGER.log(Level.FINE, () -> "Registered concurrent gauge " + metricName);
         }
@@ -377,13 +377,13 @@ public class MetricsCdiExtension implements Extension {
                                                   metric.name(), metric.absolute());
                 T instance = getReference(bm, entry.getValue().getBaseType(), entry.getKey());
                 Metadata md = Metadata.builder()
-                        .withName(metricName)
-                        .withDisplayName(metric.displayName())
-                        .withDescription(metric.description())
-                        .withType(getMetricType(instance))
-                        .withUnit(metric.unit())
-                        .reusable(false)
-                        .build();
+                    .withName(metricName)
+                    .withDisplayName(metric.displayName())
+                    .withDescription(metric.description())
+                    .withType(getMetricType(instance))
+                    .withUnit(metric.unit())
+                    .reusable(false)
+                    .build();
                 registry.register(md, instance);
             }
         });
@@ -512,13 +512,13 @@ public class MetricsCdiExtension implements Extension {
                                           bm);
                 Gauge gaugeAnnotation = site.getAnnotated().getAnnotation(Gauge.class);
                 Metadata md = Metadata.builder()
-                        .withName(gaugeID.getName())
-                        .withDisplayName(gaugeAnnotation.displayName())
-                        .withDescription(gaugeAnnotation.description())
-                        .withType(MetricType.GAUGE)
-                        .withUnit(gaugeAnnotation.unit())
-                        .reusable(false)
-                        .build();
+                    .withName(gaugeID.getName())
+                    .withDisplayName(gaugeAnnotation.displayName())
+                    .withDescription(gaugeAnnotation.description())
+                    .withType(MetricType.GAUGE)
+                    .withUnit(gaugeAnnotation.unit())
+                    .reusable(false)
+                    .build();
                 LOGGER.log(Level.FINE, () -> String.format("Registering gauge with metadata %s", md.toString()));
                 registry.register(md, dg, gaugeID.getTagsAsList().toArray(new Tag[0]));
             } catch (Throwable t) {

--- a/microprofile/metrics/src/test/java/io/helidon/microprofile/metrics/BadGaugeTest.java
+++ b/microprofile/metrics/src/test/java/io/helidon/microprofile/metrics/BadGaugeTest.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2019, 2020 Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2019, 2020 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -15,11 +15,36 @@
  */
 package io.helidon.microprofile.metrics;
 
+import java.lang.annotation.Annotation;
+import java.util.Map;
+import java.util.SortedMap;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.context.Dependent;
+import javax.enterprise.context.RequestScoped;
+import javax.enterprise.event.Observes;
+import javax.enterprise.inject.literal.SingletonLiteral;
 import javax.enterprise.inject.se.SeContainer;
 import javax.enterprise.inject.se.SeContainerInitializer;
+import javax.enterprise.inject.spi.BeanManager;
+import javax.enterprise.inject.spi.BeforeBeanDiscovery;
+import javax.enterprise.inject.spi.Extension;
 
-import org.hamcrest.Matchers;
+import io.helidon.config.mp.MpConfigSources;
+import io.helidon.microprofile.server.JaxRsCdiExtension;
+import io.helidon.microprofile.server.ServerCdiExtension;
+
+import org.eclipse.microprofile.config.Config;
+import org.eclipse.microprofile.config.spi.ConfigProviderResolver;
+import org.eclipse.microprofile.metrics.MetricID;
+import org.eclipse.microprofile.metrics.MetricRegistry;
+import org.eclipse.microprofile.metrics.MetricUnits;
+import org.eclipse.microprofile.metrics.annotation.Gauge;
 import org.jboss.weld.exceptions.DeploymentException;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Disabled;
 import org.junit.jupiter.api.Test;
 
 import static org.hamcrest.MatcherAssert.assertThat;
@@ -28,15 +53,40 @@ import static org.hamcrest.Matchers.instanceOf;
 import static org.hamcrest.Matchers.is;
 import static org.hamcrest.Matchers.notNullValue;
 import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.fail;
 
-public class BadGaugeTest  {
+public class BadGaugeTest {
+    private static final String GAUGE_NAME = "BadGaugeTest.gauge";
+    private static Config originalConfig;
+    private static ConfigProviderResolver resolver;
+    private static ClassLoader classLoader;
+
+    @BeforeAll
+    static void initClass() {
+        resolver = ConfigProviderResolver.instance();
+        classLoader = Thread.currentThread().getContextClassLoader();
+        originalConfig = resolver.getConfig(classLoader);
+
+        Config config = resolver.getBuilder()
+                .withSources(MpConfigSources.create(Map.of("server.port", "0",
+                                                           "mp.initializer.allow", "true")))
+                .build();
+        resolver.registerConfig(config, classLoader);
+    }
+
+    @AfterAll
+    static void destroyClass() {
+        if (originalConfig != null) {
+            resolver.registerConfig(originalConfig, classLoader);
+        }
+    }
 
     // TODO - remove following Disabled line once MP metrics enforces restriction
-    @org.junit.jupiter.api.Disabled
+    @Disabled
     @Test
-    public void testBadBean() {
+    void testBadBean() {
         SeContainerInitializer initializer = SeContainerInitializer.newInstance();
-        assertThat(initializer, Matchers.is(Matchers.notNullValue()));
+        assertThat(initializer, is(notNullValue()));
         initializer.addBeanClasses(BadGaugedBean.class);
         DeploymentException ex = assertThrows(DeploymentException.class, () -> {
             SeContainer cdiContainer = initializer.initialize();
@@ -54,6 +104,99 @@ public class BadGaugeTest  {
         assertThat(subCause, notNullValue());
         assertThat(subCause, is(instanceOf(IllegalArgumentException.class)));
         assertThat(subCause.getMessage(), containsString("assignment-compatible with Number"));
+    }
 
+    @Test
+    void testAppScoped() {
+        goodTest(ApplicationScoped.Literal.INSTANCE);
+    }
+
+    @Test
+    void testSingleton() {
+        goodTest(SingletonLiteral.INSTANCE);
+    }
+
+    @Test
+    void testDependentScope() {
+        // this is questionable, but it may be reporting a static field
+        goodTest(Dependent.Literal.INSTANCE);
+    }
+
+    @Test
+    void testRequestScoped() {
+        try (SeContainer ignored = startContainer(RequestScoped.Literal.INSTANCE)) {
+            // using try with resources to make sure the container is closed if it does start
+            fail("Creating the container should have thrown an exception, as request scoped beans cannot be gauge sources");
+        } catch (DeploymentException ignored) {
+            // expected
+        }
+    }
+
+    private SeContainer startContainer(Annotation scope) {
+        return SeContainerInitializer.newInstance()
+                .disableDiscovery()
+                .addExtensions(MetricsCdiExtension.class, ServerCdiExtension.class, JaxRsCdiExtension.class)
+                .addExtensions(new ScopeExtension(scope, GaugeBean.class))
+                .initialize();
+    }
+
+    private void goodTest(Annotation scope) {
+        try (SeContainer container = startContainer(scope)) {
+            MetricRegistry metricRegistry = MetricsCdiExtension.getMetricRegistry();
+            SortedMap<MetricID, org.eclipse.microprofile.metrics.Gauge> gauges = metricRegistry.getGauges();
+
+            org.eclipse.microprofile.metrics.Gauge gauge = null;
+
+            for (Map.Entry<MetricID, org.eclipse.microprofile.metrics.Gauge> entry : gauges.entrySet()) {
+                MetricID id = entry.getKey();
+                if (GAUGE_NAME.equals(id.getName())) {
+                    gauge = entry.getValue();
+                }
+            }
+
+            assertThat("Gauge should have been found", gauge, notNullValue());
+
+            assertThat(gauge.getValue(), is(7));
+
+            GaugeBean gaugeBean = container.select(GaugeBean.class).get();
+            gaugeBean.setValue(42);
+
+            if (scope == Dependent.Literal.INSTANCE) {
+                // for dependent scope, the value is reset, as we get a new instance each time
+                assertThat(gauge.getValue(), is(7));
+            } else {
+                assertThat(gauge.getValue(), is(42));
+            }
+        }
+    }
+
+    public static class GaugeBean {
+        private final AtomicInteger value = new AtomicInteger(7);
+
+        @Gauge(unit = MetricUnits.SECONDS, name = GAUGE_NAME, absolute = true)
+        public int aGauge() {
+            return value.get();
+        }
+
+        void setValue(int newValue) {
+            this.value.set(newValue);
+        }
+    }
+
+    private static class ScopeExtension implements Extension {
+        private Annotation scope;
+        private Class<?> beanClass;
+
+        private ScopeExtension(Annotation scope,
+                               Class<?> beanClass) {
+            this.scope = scope;
+            this.beanClass = beanClass;
+        }
+
+        void registerBean(@Observes BeforeBeanDiscovery bbd, BeanManager bm) {
+
+            bbd.addAnnotatedType(beanClass, beanClass.getName())
+                    .add(scope);
+        }
     }
 }

